KAFKA-17928: Make remote log manager thread-pool configs dynamic

peterxcli opened the pull request:
### Description

- [KIP-1105](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1105%3A+Make+remote+log+manager+thread-pool+configs+dynamic)
- [KAFKA-17928](https://issues.apache.org/jira/browse/KAFKA-17928?jql=project%20%3D%20KAFKA%20AND%20resolution%20%3D%20Unresolved%20AND%20assignee%20in%20(currentUser())%20ORDER%20BY%20priority%20DESC,%20updated%20DESC)

### Related Link

- [KIP-1105](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1105%3A+Make+remote+log+manager+thread-pool+configs+dynamic)
- [KAFKA-17928](https://issues.apache.org/jira/browse/KAFKA-17928?jql=project%20%3D%20KAFKA%20AND%20resolution%20%3D%20Unresolved%20AND%20assignee%20in%20(currentUser())%20ORDER%20BY%20priority%20DESC,%20updated%20DESC)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

peterxcli committed:
feat(RLM): add isReady method to RemoteLogMetadataManager

Add isReady() method to RemoteLogMetadataManager interface to check if a partition
is ready to serve remote storage operations. Implement the method across concrete
classes with appropriate checks:

- Add default implementation in RemoteLogMetadataManager interface
- Add implementation in NoOpRemoteLogMetadataManager
- Add implementation in ClassLoaderAwareRemoteLogMetadataManager
- Add implementation in TopicBasedRem

peterxcli committed:
feat(remote-log): add dynamic thread pool sizing and RLMM readiness check

Add ability to dynamically update remote storage thread pool sizes and implement
partition readiness check:

- Add methods to RemoteLogManager to update thread pool sizes:
  - updateCopyThreadPoolSize
  - updateExpirationThreadPoolSize
  - updateReaderThreadPoolSize
- Add thread pool configuration properties to DynamicRemoteLogConfig
- Add RLMM readiness check before executing remote storage tasks
- Update validation and reconfiguration logic in DynamicRemoteLogConfig

This change allows operators to tune thread pool sizes at runtime and prevents
operations on partitions that aren't ready for remote storage operations.

peterxcli committed:
Revert "feat(RLM): add isReady method to RemoteLogMetadataManager"

This reverts commit 2bbc4ad2bf03ed9ee1823256b1a13b0f4f0ef612.

peterxcli committed:
feat(remote-log): add resize method to RemoteLogManager

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1159,7 +1159,10 @@ class DynamicRemoteLogConfig(server: KafkaBroker) extends BrokerReconfigurable w
         RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPY_MAX_BYTES_PER_SECOND_PROP.equals(k) ||
-        RemoteLogManagerConfig.REMOTE_LOG_MANAGER_FETCH_MAX_BYTES_PER_SECOND_PROP.equals(k)) {
+        RemoteLogManagerConfig.REMOTE_LOG_MANAGER_FETCH_MAX_BYTES_PER_SECOND_PROP.equals(k) ||
+        RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP.equals(k) ||
```
could we move these 3 configs in a separate `if` check? This will help us in validating the change in thread count to be in the boundary (t/2 <= newSize <= t*2) similar to the validation done for other thread configs. Please refer to DynamicThreadPool.validateReconfiguration

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -289,6 +289,21 @@ public void updateFetchQuota(long quota) {
+    }
+
+    public void updateExpirationThreadPoolSize(int size) {
+        LOGGER.info("Updating remote expiration thread pool size to {}", size);
```
shall we print the before and after size in the log? similar in other places.

```
Updating remote expiration thread pool size from {} to {}
```

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -800,6 +815,11 @@ public void run() {
                     return;
                 }
 
+                if (!remoteLogMetadataManager.isReady(topicIdPartition)) {
```
Why do we need this check? Could you rebase your branch?

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1199,14 +1202,32 @@ class DynamicRemoteLogConfig(server: KafkaBroker) extends BrokerReconfigurable w
           s"old value: $oldValue, new value: $newValue")
       }
+      if (isChangedLongValue(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP)) {
+        val oldValue = oldLongValue(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP)
```
this is not a long value, it should be int. We can check directly:

```
val newRLMConfig = newConfig.remoteLogManagerConfig
val oldRLMConfig = oldConfig.remoteLogManagerConfig
if (newRLMConfig.remoteLogManagerCopierThreadPoolSize() != oldRLMConfig.remoteLogManagerCopierThreadPoolSize()) 
       ...
``

kamalcph reviewed (commented):
@peterxcli 
Thanks for the PR, could you attach unit test for this change?

peterxcli committed:
feat(remote-storage): add dynamic thread pool resizing for RemoteLogManager

Add methods to dynamically resize thread pools in RemoteLogManager:
- updateCopyThreadPoolSize
- updateExpirationThreadPoolSize
- updateReaderThreadPoolSize

The thread pool sizes can now be updated at runtime through broker configs.
Changed parameter types from int to long to match Kafka's config system.
Removed premature RLMM ready check that was causing issues.

BREAKING CHANGE: Thread pool resize methods now take long instead of int parameters

peterxcli committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17928

peterxcli committed:
refactor(remote-storage): separate thread pool and quota configurations

Split remote log manager dynamic configurations into two separate handlers:
- RemoteLogDynamicThreadPool for thread pool size configs
- DynamicRemoteLogConfig for quota and cache configs

Changes:
- Move thread pool configs from DynamicRemoteLogConfig to new RemoteLogDynamicThreadPool class
- Add thread pool configs to DynamicThreadPool.ReconfigurableConfigs
- Change thread pool size parameters from long to int
- Add thread pool size getters to KafkaConfig
- Add resize method to RemoteStorageThreadPool

This separation improves code organization by grouping related configs together
and fixes type consistency issues with thread pool sizes.

peterxcli committed:
refactor(remote-storage): rename thread pool resize methods for consistency

Rename thread pool management methods in RemoteLogManager to follow a consistent naming pattern:
- updateCopyThreadPoolSize -> resizeCopierThreadPool
- updateExpirationThreadPoolSize -> resizeExpirationThreadPool
- updateReaderThreadPoolSize -> resizeReaderThreadPool

The new names:
- Use "resize" prefix consistently with other resize operations
- Better reflect the operation being performed
- Match the naming pattern of the underlying thread pools

peterxcli committed:
test(remote-storage): add thread pool and config validation tests

Add test coverage for remote log manager configuration:
- Add thread pool resize tests to DynamicBrokerConfigTest
- Add config validation tests for remote log manager properties
- Test default values and invalid inputs for thread pools
- Test quota and timeout configuration validation

The new tests verify:
1. Dynamic thread pool resizing behavior
2. Config validation for bytes/sec quotas
3. Config validation for timeout values
4. Integration with broker config system

peterxcli commented:
Hi @kamalcph, Thanks for the review.
I do some refactors with your advice, PTAL, thanks! 😄

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -289,6 +289,21 @@ public void updateFetchQuota(long quota) {
+    public void resizeCopierThreadPool(int size) {
+        LOGGER.info("Updating remote copy thread pool size to {}", size);
+        rlmCopyThreadPool.resize(size);
+    }
```
```suggestion
    public void resizeCopierThreadPool(int newSize) {
        int currentSize = rlmCopyThreadPool.getPoolSize();
        LOGGER.info("Updating remote copy thread pool size from {} to {}", currentSize, newSize);
        rlmCopyThreadPool.resize(newSize);
   }
```

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -2163,6 +2178,10 @@ public RLMScheduledThreadPool(int poolSize, String threadPoolName, String thread
             scheduledThreadPool = createPool();
         }
 
+        public void resize(int newSize) {
```
nit: Can we rename this method to `setCorePoolSize`? Also, add one getter method `getPoolSize`

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -771,7 +772,10 @@ object DynamicThreadPool {
+    ServerConfigs.BACKGROUND_THREADS_CONFIG, 
+    RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP,
+    RemoteLogManagerConfig.REMOTE_LOG_MANAGER_EXPIRATION_THREAD_POOL_SIZE_PROP,
+    RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP)
```
We also have to update the BrokerDynamicThreadPool.reconfigurableConfigs() method to return only the below configs:

1. ServerConfigs.NUM_IO_THREADS_CONFIG
2. ReplicationConfigs.NUM_REPLICA_FETCHERS_CONFIG
3. ServerLogConfigs.NUM_RECOVERY_THREADS_PER_DATA_DIR_CONFIG and
4. ServerConfigs.BACKGROUND_THREADS_CONFIG

Otherwise for the newly added RemoteLogManager configs, it gets double registered as brokerReconfigurable in both BrokerDynamicThreadPool and  RemoteLogDynamicThreadPool.

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -797,6 +801,9 @@ object DynamicThreadPool {
       case ReplicationConfigs.NUM_REPLICA_FETCHERS_CONFIG => config.numReplicaFetchers
       case ServerLogConfigs.NUM_RECOVERY_THREADS_PER_DATA_DIR_CONFIG => config.numRecoveryThreadsPerDataDir
       case ServerConfigs.BACKGROUND_THREADS_CONFIG => config.backgroundThreads
+      case RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP => config.remoteLogCopierThreads
```
Get the copier thread size using the below way, we don't want to expose the remoteLogManager configs in KafkaConfig class:

```suggestion
      case RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP => config.remoteLogManagerConfig.remoteLogManagerCopierThreadPoolSize() 
```

same for other configs

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +847,43 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+  }
+
+  override def reconfigure(oldConfig: KafkaConfig, newConfig: KafkaConfig): Unit = {
+    val remoteLogManager = server.remoteLogManagerOpt
```
Also, please add one if check as remoteLogManger is optional:

```suggestion
   if (server.remoteLogManagerOpt.nonEmpty) {
    val remoteLogManager = server.remoteLogManagerOpt.get
    ...
    ...
```

kamalcph commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -380,6 +380,9 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
   def numIoThreads = getInt(ServerConfigs.NUM_IO_THREADS_CONFIG)
+  def remoteLogCopierThreads = getInt(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP)
+  def remoteLogExpirationThreads = getInt(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_EXPIRATION_THREAD_POOL_SIZE_PROP)
+  def remoteLogReaderThreads = getInt(RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP)
```
These getters are not required, please see above comments.

kamalcph commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1114,6 +1114,9 @@ class KafkaConfigTest {
         case RemoteLogManagerConfig.LOG_LOCAL_RETENTION_BYTES_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", -3)
+        case RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPY_MAX_BYTES_PER_SECOND_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
+        case RemoteLogManagerConfig.REMOTE_LOG_MANAGER_FETCH_MAX_BYTES_PER_SECOND_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
+        case RemoteLogManagerConfig.REMOTE_LIST_OFFSETS_REQUEST_TIMEOUT_MS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
```
this is not required.

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:
```diff
@@ -61,6 +61,10 @@ protected void afterExecute(Runnable runnable, Throwable th) {
         }
     }
 
+    public void resize(int newSize) {
```
this method is not required, we can directly call `setCorePoolSize`.

kamalcph reviewed (commented):
Thanks for the updated PR! Left few more comments to address.

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +847,43 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+      val oldValue = oldConfig.remoteLogCopierThreads
+      val newValue = newConfig.remoteLogCopierThreads
+      remoteLogManager.get.resizeCopierThreadPool(newValue)
+      info(s"Dynamic remote log manager config: ${RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPIER_THREAD_POOL_SIZE_PROP} updated, " +
```
The info log statement is not required as we already have a log inside `resizeCopierThreadPool`

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +847,43 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+      val newValue = newConfig.remoteLogExpirationThreads
+      remoteLogManager.get.resizeExpirationThreadPool(newValue)
+      info(s"Dynamic remote log manager config: ${RemoteLogManagerConfig.REMOTE_LOG_MANAGER_EXPIRATION_THREAD_POOL_SIZE_PROP} updated, " +
+        s"old value: $oldValue, new value: $newValue")
```
same as above, we can avoid double log.

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +847,43 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+      val newValue = newConfig.remoteLogReaderThreads
+      remoteLogManager.get.resizeReaderThreadPool(newValue)
+      info(s"Dynamic remote log manager config: ${RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP} updated, " +
+        s"old value: $oldValue, new value: $newValue")
```
same as above, we can avoid double log.

peterxcli committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17928

peterxcli committed:
refactor(remote-storage): improve thread pool resize logging and method signatures

- Add old size parameter to thread pool resize methods for better logging
- Rename resize() to setCorePoolSize() for clarity and consistency
- Add getCorePoolSize() method to thread pools
- Update logging to show size transitions
- Remove redundant resize() method from RemoteStorageThreadPool
- Update all callers to pass both old and new sizes

peterxcli committed:
refactor(remote-storage): simplify thread pool config logging
- Remove Logging trait from RemoteLogDynamicThreadPool
- Remove duplicate log messages from reconfigure method
- Let RemoteLogManager handle all thread pool resize logging
- Keep logging consistent and in one place

peterxcli committed:
refactor(remote-storage): move thread pool config access to RemoteLogManagerConfig
- Remove thread pool getters from KafkaConfig
- Update DynamicThreadPool to access configs via RemoteLogManagerConfig
- Update BrokerDynamicThreadPool to only handle core broker configs
- Add early return in RemoteLogDynamicThreadPool when RemoteLogManager not present
- Remove duplicate config validation tests

peterxcli committed:
chore: remove unnecessary tests for dynamic brokerconfig test

peterxcli commented:
@kamalcph Thanks for the detailed review, I made some correction base on it. 
Please take a look ~
Thanks again!

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -289,6 +289,21 @@ public void updateFetchQuota(long quota) {
         rlmFetchQuotaManager.updateQuota(new Quota(quota, true));
     }
 
+    public void resizeCopierThreadPool(int oldSize, int newSize) {
```
Instead of passing the `oldSize`, can we get it from the threadPool itself? 

```java
int currentSize = rlmCopyThreadPool.getCorePoolSize();
```

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1219,6 +1264,6 @@ object DynamicRemoteLogConfig {
     RemoteLogManagerConfig.REMOTE_LOG_MANAGER_COPY_MAX_BYTES_PER_SECOND_PROP,
     RemoteLogManagerConfig.REMOTE_LOG_MANAGER_FETCH_MAX_BYTES_PER_SECOND_PROP,
-    RemoteLogManagerConfig.REMOTE_LIST_OFFSETS_REQUEST_TIMEOUT_MS_PROP
+    RemoteLogManagerConfig.REMOTE_LIST_OFFSETS_REQUEST_TIMEOUT_MS_PROP,
```
We have to remove the trailing comma, otherwise the build will fail: 

See: https://github.com/apache/kafka/actions/runs/11892699554/job/33141799293
```
> Task :core:compileScala FAILED
```

kamalcph reviewed (commented):
Thanks for the updated PR! Left few more comments to address!

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +849,42 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+      return
+    }
+
+    val remoteLogManager = server.remoteLogManagerOpt
```
We can either rename the variable (or) get the value from the option: 

```
val remoteLogManagerOpt = server.remoteLogManagerOpt
```

(or)

```
val remoteLogManager = server.remoteLogManagerOpt.get
```

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -840,6 +849,42 @@ class BrokerDynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable
+    }
+
+    val remoteLogManager = server.remoteLogManagerOpt
+    if (newConfig.remoteLogCopierThreads != oldConfig.remoteLogCopierThreads) {
+      val oldValue = oldConfig.remoteLogCopierThreads
+      val newValue = newConfig.remoteLogCopierThreads
+      remoteLogManager.get.resizeCopierThreadPool(oldValue, newValue)
+    }
```
Better to extract the remoteLogManagerConfig: 
```
val remoteLogManager = server.remoteLogManagerOpt.get
    val oldRLMConfig = oldConfig.remoteLogManagerConfig
    val newRLMConfig = newConfig.remoteLogManagerConfig
    if (oldRLMConfig.remoteLogManagerCopierThreadPoolSize() != newRLMConfig.remoteLogManagerCopierThreadPoolSize()) {
      remoteLogManager.resizeCopierThreadPool(newRLMConfig.remoteLogManagerCopierThreadPoolSize())
    }
```

peterxcli committed:
fix: update remote log thread pool size references in DynamicBrokerConfig

- Updated references to remote log copier, expiration, and reader thread pool sizes to use the new configuration structure.
- Ensured that the resizing methods are called with the correct values from the updated configuration.

peterxcli committed:
refactor: simplify thread pool resizing methods in RemoteLogManager

- Updated the resize methods for copier, expiration, and reader thread pools to accept only the new size parameter.
- Removed the old size parameter and adjusted the logging to reflect the current size before resizing.
- Updated the DynamicBrokerConfig to call the new method signatures accordingly.

peterxcli committed:
refactor: streamline access to remote log manager configuration

- Simplified access to the remote log manager configuration by storing old and new configurations in variables.
- Removed unnecessary calls to `get` on `remoteLogManagerOpt` by directly using `get` after checking for its presence.
- Improved readability and maintainability of the code by reducing repetitive calls to configuration properties.

kamalcph commented:
@peterxcli 

There are 3 failed tests, could you fix them?
https://github.com/apache/kafka/actions/runs/11907042194/job/33180295070?pr=17793
```
FAILED ❌ KRaftClusterTest > testStartupWithNonDefaultKControllerDynamicConfiguration()
FAILED ❌ KRaftClusterTest > testIncreaseNumIoThreads()
FAILED ❌ DynamicBrokerReconfigurationTest > testThreadPoolResize(String).groupProtocol=classic
```

kamalcph commented:
@peterxcli 

I've fixed the unit tests and opened #17859 PR. Added you as the co-author. Thought to open a PR against your branch but you've multiple `trunk` merges instead of `rebase`. PTAL.

peterxcli commented:
Hi @kamalcph 
Should I close this PR?

kamalcph commented:
yes, we can close this PR and follow-up on #17859
